src/AspNetCore/Src/inprocessapplication.cxx

```diff
@@ -651,6 +652,31 @@ IN_PROCESS_APPLICATION::ExecuteApplication(
     return hr;
 }
 
+HRESULT
+IN_PROCESS_APPLICATION::OnClientDisconnect(
+    IHttpContext* pHttpContext,
+    BOOL fClientDisconnected
+)
+{
+    HRESULT hr;
+    IN_PROCESS_STORED_CONTEXT* pInProcessStoredContext = NULL;
+
+    if ( m_ClientDisconnectHandler == NULL || !fClientDisconnected)
+    {
+        return E_FAIL;
+    }
+
+    hr = IN_PROCESS_STORED_CONTEXT::GetInProcessStoredContext( pHttpContext, &pInProcessStoredContext );
+    if ( FAILED( hr ) )
+    {
+        // Finish the request as we couldn't get the callback
+        pHttpContext->GetResponse()->SetStatus( 500, "Internal Server Error", 19, hr );
```

Comment:
Most likely this function will be called ion another thread. Setting response status in failure case will not take effect as the main thread still process the request and will set http response. I think this function can be simplified.
We also need some kind interlockexchange mechanism to prevent main thread from the race condition. I would suggest to process against forwarding handler. That is main thread and the callback thread should set/check a flag in forwarding handler to hold the stored context.

Reply:
@pan-wang what thread does this fire on? Some random http.sys thread?

Reply:
Some callback thread from HTTP.SYS. this is why we need a lock/flag mechanism to ensure the httpcontext is still valid.  For example,  before finish the request,  we need unregistered the callback and set a flag in forwarding handler.  The disconnect callback need to check the flag and set it (if it was not set yet) to prevent main thread release the request/forwarding handler object. This is a very dangerous code path which can easily cause AV and thus security issue.

Reply:
Agreed. Do we have similar issues with the out of process code path as well?